src/node/ledger_secrets.h

```diff
@@ -49,154 +46,107 @@ namespace ccf
-    std::map<kv::Version, LedgerSecret> secrets_map;
+    };
+    // List of ledger secrets that are valid from a specific version to the
+    // version of the next entry in the list. The last entry in the list is
+    // valid for all subsequent versions.
+    std::list<VersionedLedgerSecret> secrets_list;
 
     LedgerSecrets() = default;
 
-    // Called on startup to generate fresh ledger secret
-    LedgerSecrets(std::shared_ptr<AbstractSeal> seal_, bool force_seal = true) :
-      seal(seal_)
-    {
-      // Generate fresh ledger encryption key
-      add_secret(1, LedgerSecret(true), force_seal);
-    }
+    LedgerSecrets(std::shared_ptr<AbstractSeal> seal_) : seal(seal_) {}
 
     // Called when a node joins the network and get given the ledger secrets
     // since the beginning of time
     LedgerSecrets(
       LedgerSecrets&& ledger_secrets_, std::shared_ptr<AbstractSeal> seal_) :
-      secrets_map(std::move(ledger_secrets_.secrets_map)),
+      secrets_list(std::move(ledger_secrets_.secrets_list)),
       seal(seal_)
     {}
 
     bool operator==(const LedgerSecrets& other) const
     {
-      return secrets_map == other.secrets_map;
+      return secrets_list == other.secrets_list;
     }
 
-    // Called when a backup is given past ledger secret via the store
-    bool set_secret(kv::Version v, const std::vector<uint8_t>& raw_secret)
+    void init(kv::Version version = 1)
     {
-      auto search = secrets_map.find(v);
-      if (search != secrets_map.end())
+      if (secrets_list.size() != 0)
       {
-        return false;
+        throw std::logic_error("Ledger secrets have already been initialised!");
       }
 
-      add_secret(v, LedgerSecret(raw_secret), false);
-
-      return true;
+      secrets_list.push_back({version, LedgerSecret()});
     }
 
-    // Called when sealed secrets need to be stored during recovery
-    std::vector<kv::Version> restore(const nlohmann::json& sealed_secrets)
+    LedgerSecret get_latest()
     {
-      std::vector<kv::Version> restored_versions;
+      return secrets_list.back().secret;
+    }
 
-      for (auto it = sealed_secrets.begin(); it != sealed_secrets.end(); ++it)
+    std::optional<LedgerSecret> get_penultimate()
+    {
+      if (secrets_list.size() <= 1)
       {
-        kv::Version v = std::stoi(it.key());
-
-        auto search = secrets_map.find(v);
-        if (search != secrets_map.end())
-        {
-          throw std::logic_error(fmt::format(
-            "Cannot restore ledger secret that already exists: ", v));
-        }
-
-        auto s = seal->unseal(it.value());
-        if (!s.has_value())
-        {
-          throw std::logic_error(
-            fmt::format("Ledger secret could not be unsealed: {}", v));
-        }
-
-        LOG_DEBUG_FMT(
-          "Ledger secret successfully unsealed at version {}", it.key());
-
-        add_secret(v, LedgerSecret(s.value()), false);
-
-        restored_versions.push_back(v);
+        return {};
```

Comment:
Minor: Please `return std::nullopt;` here. I know this does the correct thing, but I'm wary of the obscure bugs when the return type changes and we miss that this does something totally different.

Reply:
Since this is minor, I will fix this in the follow-up PR which will come later today.